storm-core/src/clj/backtype/storm/stats.clj

```diff
@@ -373,4 +360,1248 @@
+     :acked (sum (vals sid->num-acked))}))
+
+(defn add-pairs
+  ([] [0 0])
+  ([[a1 a2] [b1 b2]]
+   [(+ a1 b1) (+ a2 b2)]))
+
+(defn mk-include-sys-fn
+  [include-sys?]
+  (if include-sys?
+    (fn [_] true)
+    (fn [stream] (and (string? stream) (not (system-id? stream))))))
+
+(defn mk-include-sys-filter
+  "Returns a function that includes or excludes map entries whose keys are
+  system ids."
+  [include-sys?]
+  (if include-sys?
+    identity
+    (partial filter-key (mk-include-sys-fn false))))
+
+(defn- agg-bolt-streams-lat-and-count
+  "Aggregates number executed and process & execute latencies."
+  [idk->exec-avg idk->proc-avg idk->executed]
+  {:pre (apply = (map #(set (keys %))
+                      [idk->exec-avg
+                       idk->proc-avg
+                       idk->executed]))}
+  (letfn [(weight-avg [id avg] (let [num-e (idk->executed id)]
+                                   (if (and avg num-e)
+                                     (* avg num-e)
+                                     0)))]
+    (into {}
+      (for [k (keys idk->exec-avg)]
+        [k {:executeLatencyTotal (weight-avg k (idk->exec-avg k))
+            :processLatencyTotal (weight-avg k (idk->proc-avg k))
+            :executed (idk->executed k)}]))))
+
+(defn- agg-spout-streams-lat-and-count
+  "Aggregates number acked and complete latencies."
+  [idk->comp-avg idk->acked]
+  {:pre (apply = (map #(set (keys %))
+                      [idk->comp-avg
+                       idk->acked]))}
+  (letfn [(weight-avg [id avg] (let [num-e (get idk->acked id)]
+                                   (if (and avg num-e)
+                                     (* avg num-e)
+                                     0)))]
+    (into {}
+      (for [k (keys idk->comp-avg)]
+        [k {:completeLatencyTotal (weight-avg k (get idk->comp-avg k))
+            :acked (get idk->acked k)}]))))
+
+(defn swap-map-order
+  "{:a {:A 3, :B 5}, :b {:A 1, :B 2}}
+    -> {:A {:b 1, :a 3}, :B {:b 2, :a 5}}"
+  [m]
+  (apply merge-with
+         merge
+         (map (fn [[k v]]
+                (into {}
+                      (for [[k2 v2] v]
+                        [k2 {k v2}])))
+              m)))
+
+(defn- compute-agg-capacity
+  "Computes the capacity metric for one executor given its heartbeat data and
+  uptime."
+  [m uptime]
+  (when uptime
+    (->>
+      ;; For each stream, create weighted averages and counts.
+      (merge-with (fn weighted-avg+count-fn
+                    [avg cnt]
+                    [(* avg cnt) cnt])
+                  (get (:execute-latencies m) "600")
+                  (get (:executed m) "600"))
+      vals ;; Ignore the stream ids.
+      (reduce add-pairs
+              [0. 0]) ;; Combine weighted averages and counts.
+      ((fn [[weighted-avg cnt]]
+        (div weighted-avg (* 1000 (min uptime 600))))))))
+
+(defn- page+comp-dispatch
+  [page-type comp-type & _]
+  [page-type comp-type])
+
+(defmulti agg-pre-merge
```

Comment:
I'm not sure a defmulti is the cleanest here.  All the places it is used are hard code the first two parameters which are used for routing.